### ISSUE TITLE
Tracking Issue for RFC 213: Default Type Parameter Fallback

### ISSUE THREAD

**Issue by jroesch** [rust-lang/rust#27336]
2024-07-15: This feature (`default_type_parameter_fallback`) is slated for removal, but that is currently blocked on too many [crater regressions](https://github.com/rust-lang/rust/pull/127655#issuecomment-2228285460).

-----

EDIT: this issue has been stalled on disagreements about how to handle a nasty problem found during implementation. [See the internals thread where this was detailed and discussed](https://internals.rust-lang.org/t/interaction-of-user-defined-and-integral-fallbacks-with-inference/2496).

-----

This is a tracking issue for [RFC 213](https://github.com/rust-lang/rfcs/blob/master/text/0213-defaulted-type-params.md). 

The initial implementation of this feature has [landed](https://github.com/rust-lang/rust/pull/26870).

cc @nikomatsakis

**Comment by bluss** [comment]
What is the status of this?

**Comment by pnkfelix** [comment]
@bluss it is still feature-gated, AFAICT.  See e.g. the discussion on PR #26870 , or just look at this [playpen](http://is.gd/RcQ8LC)

I am not sure what the planned schedule is for unfeature-gating it.

nominating for discussion.

**Comment by mahkoh** [comment]
This doesn't seem to work to a useful degree.

```rust
#![crate_type = "lib"]
#![feature(default_type_parameter_fallback)]

trait A<T = Self> {
    fn a(t: &T) -> Self;
}

trait B<T = Self> {
    fn b(&self) -> T;
}

impl<U, T = U> B<T> for U
    where T: A<U>
{
    fn b(&self) -> T {
        T::a(self)
    }
}

struct X(u8);

impl A for X {
    fn a(x: &X) -> X {
        X(x.0)
    }
}

fn f(x: &X) {
    x.b(); // ok
}

fn g(x: &X) {
    let x = x.b();
    x.0; // error: the type of this value must be known in this context
}
```

**Comment by jroesch** [comment]
@mahkoh there is a necessary patch that hasn't gotten rebased since I stopped my summer internship. I've been unfortunately busy with real life stuff, looks like @nikomatsakis has plans for landing a slightly different version according to a recent post of his on the corresponding documentation issue for this feature.

**Comment by bluss** [comment]
@nikomatsakis I know the lang team didn't see any future in this proposal, will you put that on record in the issue :smile:?

One example where this seems to be the only way out is this concrete example of API evolution in libstd.

`Option<T>` implements `PartialEq` today, but we would like to extend it to `PartialEq<Option<U>> where T: PartialEq<U>`. It appears this feature can solve the type inference regressions that would otherwise occur (and might block us from doing this oft-requested improvement of Option).

**Comment by nikomatsakis** [comment]
@bluss I HAVE been dubious of this feature, but I'm been slowly reconsidering. @aturon is supposed to be doing some exploration of this whole space and writing up some detailed thoughts. I actually started rebasing @jroesch's dead branch to implement the desired semantics and making some progress there too, but I've been distracted. 

One advantage of finishing up the impl is that it would let us experiment with extensions like the one you describe to see how backwards compatible they truly are -- one problem with fallback is that it is not ACTUALLY backwards compatible, because of the possibility of competing incompatible fallbacks.

**Comment by nikomatsakis** [comment]
That said I still have my doubts :)

**Comment by nikomatsakis** [comment]
Another example where this could be useful -- basically the same example as petgraph -- is adding allocators to collections in some smooth way.

**Comment by durka** [comment]
What are the drawbacks to turning this on? It seems to mainly make things compile that otherwise cannot infer enough type information.

**Comment by cybergeek94** [comment]
I have a pretty good use for this too. It's basically what @bluss mentioned, adding new types to an `impl` without breaking inference on existing usage.

**Comment by withoutboats** [comment]
Is the only issue with this the interaction with numeric fallback? I like default type parameters a lot.

**Comment by joshtriplett** [comment]
I have a use case for this feature as well.  Consider the following code:

```rust
#![feature(default_type_parameter_fallback)]
use std::path::Path;

fn func<P: AsRef<Path> = String>(p: Option<P>) {
    match p {
        None => { println!("None"); }
        Some(path) => { println!("{:?}", path.as_ref()); }
    }
}

fn main() {
    func(None);
}
```

Without `default_type_parameter_fallback`, the call in `main` would require a type annotation: `func(None::<String>);`.

Along similar lines, consider a function accepting an `IntoIterator<Item=P>`, where `P: AsRef<Path>`.  If you want to pass `iter::empty()`, you have to give it an explicit type.  With `default_type_parameter_fallback`, you can just pass `iter::empty()`.

**Comment by kennytm** [comment]
Note: In case anyone hit the "type macros are experimental" error in 1.8.0, the `type_macros` feature is tracked at #27245 (the wrong number has been corrected in 1.9.0 with #32516).

**Comment by tikue** [comment]
I've found this feature immensely helpful and would love to see it stabilized.

**Comment by Ericson2314** [comment]
I think this feature is very useful for an ergonomic Read/Write with an associated error type. See https://github.com/RustOS-Fork-Holding-Ground/coreio/commit/81212f52e0103272ae7ecacc1818a3a50d36dbb1 for how this was formerly done.

**Comment by withoutboats** [comment]
If we had this feature, possibly we could have slice::sort take a type parameter to provide the sorting algorithm.

**Comment by nikomatsakis** [comment]
So @eddyb floated an interesting idea for how to have this feature without the forwards compatibility hazards. I'm not honestly sure how much what has been written about and where -- but in general there is an obvious problem when you introduce "fallback" that it could happen that a given type variable has multiple fallbacks which apply. This means that introducing a new type parameter with a fallback can easily still be a breaking change, no matter what else we do. This is (one of) the reasons that my enthusiasm for this feature has dulled a bit.

You can see this problem immediately when you consider the interaction with the `i32` fallback we have for integers. Imagine you have `foo(22)` where the function `foo` is defined as `fn foo<T>(t: T)`. Currently `T` will be `i32`. But if you change `foo` to `fn foo<T=u32>(t: T)`, then what should `T` be? There are now two potentially applicable defaults: i32 and u32.

The idea that @eddyb had was basically to be *more* conservative around defaults. In particular, the idea was that we would limit defaults to type declarations (iirc) and not to fns. I'm having trouble recalling the precise plan he put forward -- it was quickly and over IRC -- but iirc the idea was that it would be an error if to have a type variable that had no default mix with one that had some other default. All the type variables would have to have the same default. 

So e.g. you could add an allocator parameter `A` to various data-structures like `Vec` and `HashMap`:

```rust
struct Vec<T, A=GlobalAlocator> { ... }
struct HashMap<K, V, A=GlobalAllocator> { ... }
````

so long as you are consistent about using the **same** default for allocators in every other place that you add them, since otherwise you risk having defaults that disagree. Don't have time to do a detailed write-up, and I'm probably getting something a bit wrong. Perhaps @eddyb can explain it better.

**Comment by eddyb** [comment]
The gist of the idea is that *before* applying defaults, we ensure that everything which *could* have defaults added in the future was *already* inferred (i.e. as we error now for unbound inference variables).

So if you had `struct Foo<T, A>(Box<T, A>);` (and `Box` had a default for `A`), `let _ = Foo(box x);` would *always* be an error, as the result of inference would change if a default for `A` was added to `Foo`.

Your options would be `struct Foo<T>(Box<T>);` or `struct Foo<T, A=GlobalAllocator>(Box<T, A>);`: any other default would be useless during inference because it would conflict with `Box`'s default.

This scheme works with both allocators and hashers AFAICT, since you can just use the "global default" everywhere you want to make it configurable, and there are likely more usecases out there like that.

The catch is that you have to limit from the start the possible locations of defaults you'll take into consideration to "complete" type inference, and allowing them on more than type definitions would require a lot of duplication of the same default everywhere, but there may be a reason to do that.

**Comment by eddyb** [comment]
@nikomatsakis We could "inherit" defaults from `Self`'s type parameters in inherent `impl`s and allow defaults everywhere else, what do you think?
It seems rare that you have fully parametric free `fn`s working with a container type that wants defaults.
We could also "inherit" defaults everywhere, from user type definitions and forbid having your own defaults for type parameters that end up being used in types which could have defaults in the future.

Such an adjustment would make this viable even in @withoutboats' `<[T]>::sort` situation.

**Comment by withoutboats** [comment]
> You can see this problem immediately when you consider the interaction with the i32 fallback we have for integers. Imagine you have foo(22) where the function foo is defined as fn foo<T>(t: T). Currently T will be i32. But if you change foo to fn foo<T=u32>(t: T), then what should T be? There are now two potentially applicable defaults: i32 and u32.

I don't know, it seems to me like adding a default to an existing type parameter just ought to be a breaking change?

**Comment by joshtriplett** [comment]
Ideally, it should be possible to go from a specific type to a type parameter with a default without a breaking change. That would allow generalizing a function.

**Comment by eddyb** [comment]
@withoutboats The problem is that then the effects on existing code can be silent and unpredictable, and that is the reason progress on this issue has been stalled.
And if adding defaults is a breaking change, we can't use them for API evolution.

**Comment by withoutboats** [comment]
@eddyb @joshtriplett I don't see how going from a type to a type parameter with a default to that type could be a breaking change. I'm saying that going from a type parameter without a default to a parameter without a default should just be defined as a breaking change. That is:

```rust
// Not breaking, as far as I can tell

// From:
fn foo(x: u32) { }
// To:
fn foo<T=u32>(x: T) { }

// Breaking

// From:
fn foo<T>(x: T) { }

// To:
fn foo<T=u32>(x: T) { }
```

**Comment by joshtriplett** [comment]
@withoutboats I can imagine cases where going from a concrete type to a type parameter could cause exactly the same breakage.

Consider the following code:

```rust
fn foo(x: SomeOtherType) { ... }
fn bar<T=SomeType>() -> T { ... }

fn main() {
    let x = bar();
    foo(x);
}
```

That code should compile, inferring `x` to have type `SomeOtherType`, and thus inferring bar's `T` as `SomeOtherType`.  However, if you change `foo` to:

```rust
fn foo<T=SomeOtherType>(x: T) { ... }
```

then that code will fail to compile, because it can't unambiguously infer the type of `x`.

A common real-world example of this would involve various combinations of OsString, Path, and similar string types, and attempting to generalize a function that takes one of those types.  For instance, I'd like to be able to generalize a function that takes a `&Path` to a function that takes an `AsRef<Path>` and defaults to `&Path`.  But this *could* generate breakage and type inference failures.

**Comment by eddyb** [comment]
@withoutboats The concrete type could have unified with an unlimited number of type variables that now remain unbound. If any of them have defaults, they can now conflict, and no universal priority ordering can really exist. If they don't, then they're not future-proof.

Enabling defaulting for everyone has the same effect as adding a type parameter with a default (e.g. for `HashMap`'s hasher), although I suppose that still doesn't require that adding a default to an existing type parameter is not a breaking change.

**Comment by withoutboats** [comment]
@eddyb @joshtriplett That makes sense, thanks.

**Comment by eddyb** [comment]
@withoutboats Yes, that is a very fortunate case of a type parameter that can only be explicitly provided or inferred as its default, not bound to anything else.

**Comment by jethrogb** [comment]
Another use case: I'd really like to move HashMap/HashSet to libcollections from libstd. The problem is `RandomState` and in particular `HashMap<K,V,RandomState>::new`. RandomState needs to live in libstd, but you can't split the impl like that between libcollections and libstd. So one could try to have a generic implementation in libcollections and a type alias in libstd like `type HashMap<K,V,S=RandomState> = core_collections::HashMap<K,V,S>`  combined with a impl of `HashMap<K,V,BuildHasher+Default>::new` in libcollections. This however results in the exact inference problem that this feature is designed to solve.

**Comment by nrc** [comment]
@eddyb to clarify some points about your suggestion:

* it must be enforced at usage sites, not in definitions? In your example, it is `let _ = Foo(box x)` that is the problem, specifically if the type of `x` cannot be fully inferred?
* do you intend to forbid defaults for functions entirely? Could you give an example of functions with defaults which are more problematic than types?
* could you explain why using Self's types makes the sort example work?
* adding a default to an existing type parameter could still be a breaking change?

**Comment by eddyb** [comment]
> * it must be enforced at usage sites, not in definitions? In your example, it is `let _ = Foo(box x)` that is the problem, specifically if the type of `x` cannot be fully inferred?

Yes, however, the enforcement on uses is more permissive than today's type-checking, i.e. it would only allow more code to compile. And it's not the type of `x`, but the fact that `box x` could have any `A`, and the lack of a default on `Foo` represents a hazard.

> * do you intend to forbid defaults for functions entirely? Could you give an example of functions with defaults which are more problematic than types?

It's not functions *with* defaults, it's functions with *any* type parameters that would cause problems in such as scheme. Without their own defaults, matching everything else (or automatically deduced), they would prevent *any* inference variable they come into contact from getting its defaults from *anywhere*.

> * could you explain why using Self's types makes the sort example work?

Not just `Self`, but *everything*. The sort example would be helped by taking into account defaults of type parameters of types used in the signature *and* the fact that the type parameter would *not* be used in the signature, e.g.:
```rust
impl<T> [T] {
    pub fn sort<A: Algorithm = MergeSort>(&mut self) {...}
}
```
In such a definition, `A` can either be explicitly provided *or* defaulted to `MergeSort`.
It's *very* important that `A` *can't* be inferred from *anywhere* else, which results in 0 hazards.

> * adding a default to an existing type parameter could still be a breaking change?

I don't see how. All the code that had compiled without the default succeeded in inference *without* applying *any* defaults, so it couldn't *ever* see the new default.

**Comment by nrc** [comment]
Thanks for the explanations!

> It's not functions with defaults, it's functions with any type parameters that would cause problems in such as scheme. Without their own defaults, matching everything else (or automatically deduced), they would prevent any inference variable they come into contact from getting its defaults from anywhere.

So, we apply the same rules to type parameters on both functions and types?

And to summarise your rule, is it accurate to say "wherever we perform inference, if an inference variable has a default, then it is an error if that variable unifies with any other inference variables, unless they have the same default or the variable also unifies with a concrete type"?

**Comment by eddyb** [comment]
@nrc Well, we *could* apply the same rules, we could deny/ignore defaults on functions, or we could make functions go somewhere in between, using defaults that we can gather from their signature, combined with their own defaults, in cases where they would be unambiguous.
I tried to list the options, and I prefer the hybrid rule for functions, but it's just one option of several.

As for your formulation, if you apply that rule after inference stops at a fix-point, it would amount in "for all remaining unassigned inferences variables, first error for any without a default" (because they'd never unify with anything that's not caused by applying a default), "and then apply all defaults at the same time" (with conflicting defaults causing type mismatch errors).
So yes, I believe that is correct, it seems to be equivalent to the algorithm I had in mind.

*However*, if you try to apply defaults *before* inference has given up, or your "unifies with a concrete type" can use types that were a side-effect of applying defaults, the compatibility hazards remain.

**Comment by jsen-** [comment]
I have this convenient command executor which deals with all the things that might go wrong including checking the exit code.
Obviously in 99% of cases non-zero exit code means an error, but I'd like to cover this 1% without resorting to always requiring some exitcode validator.

I'm reading that `default_type_parameter_fallback` is not likely to be stabilized, is there some other way to model those 99/1 use cases without resorting to something like `fn execute_command_with_custom_exit_code_validator(...)`?

```rust
pub trait ExitCodeValidator {
    fn validate(exit_code: i32) -> bool;
}

pub struct DefaultExitCodeValidator;

impl ExitCodeValidator for DefaultExitCodeValidator {
    fn validate(exit_code: i32) -> bool {
        return exit_code == 0;
    }
}

pub fn execute_command<T, V=DefaultExitCodeValidator>(cmd: &mut Command) -> Result<T>
where V: ExitCodeValidator,
{
    let output = ...;
    let exit_code = ...;

    if V::validate(exit_code) {
        Ok(output)
    } else {
        Err(InvalidExitCode(exit_code))
    }
}
```

**Comment by cramertj** [comment]
Correct me if I'm wrong, but the behavior being discussed here seems to be available on both stable and nightly (without a warning). Even though defaulted type parameters for functions are feature-flagged, it's possible to emulate this behavior by making a function impl on a struct with a default type parameter.

This behavior can be seen [here](https://is.gd/KILhJr). Am I mistaken in thinking that this is exactly the behavior this conversation is intending to prevent? I've changed i8 from a concrete type to a default, and it resulted in a change of the type of the resulting value.

**Comment by jsen-** [comment]
I see. I guess this does not clutter the API that much and still monomorphises the call, so you don't have to pay for indirection as with builder pattern alternative. Cool, thanks for pointing this out.

**Comment by cramertj** [comment]
@jsen- I think you misunderstood my point. I had thought I had discovered a bug-- default type parameters *shouldn't* be usable on functions (currently). However, I was just mistaking numeric defaults for something more sinister.

**Comment by nikomatsakis** [comment]
@jsen- it doesn't directly help you, but it occurs to me that if we had default fn parameters, one could (maybe?) write:

```rust
pub trait ExitCodeValidator {
    fn validate(&self, exit_code: i32) -> bool;
}

pub struct DefaultExitCodeValidator;

impl ExitCodeValidator for DefaultExitCodeValidator {
    fn validate(&self, exit_code: i32) -> bool {
        return exit_code == 0;
    }
}

pub fn execute_command<T, V>(
    cmd: &mut Command,
    validator: V = DefaultExitCodeValidator)
    -> Result<T>
where V: ExitCodeValidator,
{
    let output = ...;
    let exit_code = ...;

    if validator.validate(exit_code) {
        Ok(output)
    } else {
        Err(InvalidExitCode(exit_code))
    }
}
```

This seems better than your existing code because it allows an exit code validator to carry some state. But of course it relies on a future that's not implemented, and in particular on the ability to instantiate a default parameter of generic type.

**Comment by jsen-** [comment]
@nikomatsakis right, I just shortened my current implementation, which relies on `default_type_parameter_fallback`.
I'd sure love to see default function arguments, but I'm not even aware of any plans to include them in the language (I'll sure look around in the RFC section 😃)
I posted the question because we need to move to `stable` some day 😉

@cramertj yeah, I was confused. Based on your comment I got an idea, that I could emulate it with *struct* defaulted type params.
Something like this:
```rust
// defaulted:
Exec::execute_command(cmd);
// explicit:
Exec::<MyExitCodeValidator>::execute_command(cmd);
```
...but I'm finding this is not be possible as well. I'm not yet fully shifted away from C++'s meta-programming. Definitely still need to learn a lot about rust.

**Comment by eddyb** [comment]
@jsen- Try `<Exec>::execute_command(cmd)` (that way the path is in a type context, not expression context, and the defaults are forcefully applied).

**Comment by jsen-** [comment]
@eddyb thanks a lot, that worked perfectly
Here's a link to somewhat contrived, but working [example](https://is.gd/MvXOiG) if that helps anyone who finds this

**Comment by nikomatsakis** [comment]
adding a link to this (old) internals thread for posterity:

https://internals.rust-lang.org/t/interaction-of-user-defined-and-integral-fallbacks-with-inference/2496

it discusses the interaction of RFC 213 with integral fallback, though similar issues can also arise without using integers, if you just have multiple functions with competing fallback.

**Comment by Gankro** [comment]
It's really unfortunate that this is blocking #32838. Is there anything I can do to help push this forward, outside of implementation work?

**Comment by Ericson2314** [comment]
I don't think it's blocking that. We can always just newtype all of collections in std, and move hashmap into collections while we are at it.

It's an ugly solution, but only a temporary once. The allocator traits are far enough behind schedule that we should seriously consider it.

**Comment by nikomatsakis** [comment]
@Gankro I think we're still sort of in need of a good survey of the space of ideas and a nice summary. That would be helpful.

**Comment by mitsuhiko** [comment]
I have a few more cases where this is really useful. In particular I have a situation where I have a generic function that requires type inference for the return value. However if there is no user of the return value it requires annotations even though I would have perfectly sensible defaults that then get optimized away.

Think of it like this:

```rust
// currently required
let _: () = my_generic_function()?;

// with a unit default
my_generic_function()?;
```

**Comment by leodasvacas** [comment]
I'm no expert but I'll try to summarize and contribute.

We want to use type defaults to inform inference. However we may get multiple defaults that could apply to an inference variable, resulting in a conflict. Also new defaults may be added in the future, causing conflicts where there was none.

@eddyb proposes the conservative approach of erroing on conflict and erroing when trying to apply a default to a type with no default, for future-proofing. The consequence is that all defaults must be present and must agree to be applied.

Let's take the reasonable and useful [example](https://github.com/rust-lang/rust/issues/27336#issuecomment-206116131) by @joshtriplett. It would not work because it's trying to unify `T` in `struct Option<T>` with `P` in `fn func<P: AsRef<Path> = String>(p: Option<P>)`. Even if `Option<T>` were to gain a default for `T`, intuitively `func` should not care because it's default is _more local_ and should take precedence.

So I propose that _local defaults trump type defaults_. So fn and impl defaults would take precedence over the default on the type, meaning that any future conflict between `T` and `P` above would always be solved by using the default on `P`, making it future-proof to apply `P`'s default which is `String`.

Note that this is limited to literals, if the value came from another fn as in:
```rust
fn noner<T>() -> Option<T> { None }

fn main() {
    // func is same as in original example.
    func(noner());
}
```
Then we are back to an error.

This a small yet useful extension, together with the `slice::sort` example this makes a case for allowing type defaults in impls and fns. Hopefully we are coming to a _minimal but useful_ version of this feature that we could stabilize.

As extensions we could consider how to make the ordering among defaults more fine-grained. For example giving lower priority to types that appear on return types since those are likely inferred from latter usage, solving the `noner` case and constructors in general.

**Comment by DoumanAsh** [comment]
Found this issue when I was looking into RFC.

I noticed recently a problem for myself when I have default type paramer, but my impl is generic.
So currently I have two options:
- impl for struct with default type parameter
- Ask users to use a bit ugly `<Struct>` syntax to guarantee default type paramer.

I wonder if it would be sensible to allow impl blocks to have default type parameters while retaining ability to specify trait boundry.

e.g. 

```rust
pub trait MyTrait {}

struct MyStruct;
impl MyTrait for MyStruct;

struct API<M: MyStruct> {
....
}

impl<M: MyTrait=MyStruct> for API<M> {
    pub fn new() -> Self<M> {
    }
}
```

I may understand reasoning for impl blocks to overshadow default parameter when invoking `API::new()` but it is a bit sad when you cannot clearly set default type parameter for impl block alongside your struct.

**Comment by lachlansneff** [comment]
Could this be fast tracked? Having better type inference would allow better user-experience for collections with custom allocators. https://github.com/rust-lang/rust/issues/42774#issuecomment-464367839

**Comment by Gankro** [comment]
The proposal is in currently in a zombie state due to disagreements among the lang-team over how to resolve implementation issues. No progress has been made as it is otherwise regarded as low priority.

**Comment by jethrogb** [comment]
Maybe that priority should be re-evaluated? I think this is blocking various library improvements such as `std`-compatible HashMap in `alloc`, custom allocators, etc.

**Comment by Gankro** [comment]
I'm not aware of any reason custom allocators would be blocks on default-affects-inference. They can be added backwards-compatibly in the same way hashers currently work -- ::new and friends would hardcode the default, and Default::default/with_allocator would be properly generic.

**Comment by npmccallum** [comment]
Is this the same issue I'm running into here?

```rust
pub struct Foo<T: Default = u64>(T);

impl<T: Default> Foo<T> {
    pub fn make() -> Foo<T> {
        Foo(T::default())
    }
}

fn main() {
    // error[E0283]: type annotations required: cannot resolve `_: std::default::Default`
    let foo = Foo::make();
}
```

**Comment by Gankro** [comment]
Yes.

**Comment by Ericson2314** [comment]
@Gankro 

> I'm  not aware of any reason custom allocators would be blocked on  default-affects-inference. They can be added backwards-compatibly in the  same way hashers currently work -- ::new and friends would hardcode the  default, and Default::default/with_allocator would be properly generic.

The default hasher `RandomState` for `HashMap` cannot go in `alloc` because it accesses os-specific sources of randomness. (https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#3165-3186) We therefore need to define the default "after the fact". https://github.com/rust-lang/rfcs/pull/2492 is the only way I know to get around this.

I'd also like to define collections prior to the notion of global allocation being introduced, which would piggyback on the solution for `RandomState`. This, of course, is not a blocker.

**Comment by ethindp** [comment]
So, I don't really chip in (I'm not good with language design), but for the `RandomState` issue could we just use the RDRAND/RDSEED CPU instructions when no OS is available to supply a source of randomness? You could evn have a panic (or possibly compile-time check) to determine if those instructions are supported and, if not, either fail with an error or panic.

**Comment by joshtriplett** [comment]
What is the current status of this feature? What issues would need to be resolved to make this possible to stabilize?

The internals thread linked from the top post of this issue, regarding interaction with integer literal inference, seems to have come to a conclusion. What other blockers exist?

**Comment by nikomatsakis** [comment]
My take is that this feature is sort of back to "square one". I don't remember how much of the implementation exists, but I at least still have some strong concerns about "open ended" type parameter fallback. I'm not opposed to somebody picking up and trying to explore the issue again, but I would want to kind of "start over" by reviewing the use cases -- as concrete as possible! -- and the 'palette' of solutions available.

**Comment by nikomatsakis** [comment]
I think we did rip out some of the code related to this at some point, as well.

**Comment by Gankra** [comment]
I think we should "unaccept" the RFC to make it clear this needs to be redesigned from the ground up, go through RFCs again.

**Comment by nikomatsakis** [comment]
I would be in favor of that, and removing whatever code remains. Sort of like @Mark-Simulacrum did for the `x <- y` operator.

**Comment by Mark-Simulacrum** [comment]
I think that was actually @aidanhs :)

I would personally also be in favor of closing tracking issues in favor of new RFCs where there is significant design work remaining and the work is not actively in cache for some group of people.

**Comment by nikomatsakis** [comment]
Yes, you're right, sorry @aidanhs =)

**Comment by XAMPPRocky** [comment]
I'm going to close this since there seems to be a agreement to close above and there hasn't been any activity since.

**Comment by joshtriplett** [comment]
For anyone following this who may wish to start a new effort to solve it (which should probably begin with a lang MCP), one of the biggest use cases is adding a new type parameter to an existing type, without breaking semver compatibility.

**Comment by varkor** [comment]
> one of the biggest use cases is adding a new type parameter to an existing type, without breaking semver compatibility.

Presumably there's some overlap here with https://github.com/rust-lang/wg-allocators/issues/2, though I haven't been following this issue closely.

**Comment by wyfo** [comment]
I've recently noticed the following "hack":
```rust
enum Foo<T=String> {
    Bar(T),
    Baz,
}

fn main() {
    // let _ = Foo::Baz; // Doesn't compile
    let _: Foo = Foo::Baz; // Compiles, as noticed in #95486
    let _ = <Foo>::Baz; // Compiles fine too!
}
```
`<Foo>::Baz` is also more convenient, as it can be used directly as an expression. I did believe that `T::` and `<T>::` were equivalent, but it's not the case.

**Comment by Mitiko** [comment]
> I did believe that `T::` and `<T>::` were equivalent, but it's not the case.

What's the difference between `T::` and `<T>::`?
Can somebody explain to me why this hack actually works?

**Comment by QuineDot** [comment]
> > I did believe that `T::` and `<T>::` were equivalent, but it's not the case.
> 
> What's the difference between `T::` and `<T>::`? Can somebody explain to me why this hack actually works?

Types in part of a (path) expression, like `T::`, use inference for all elided type parameters -- including defaulted ones -- and do not fallback to the defaulted type when running into ambiguity.  That's pretty much what this RFC was about.  So these are equivalent and cannot currently fallback to the defaulted type (and thus fail to compile):
```
    let _ = Foo::Baz;
    let _ = Foo::<_>::Baz;
```
Where as in type ascription location, you can't elide non-defaulted parameters, but if you elide defaulted parameters they take on the default value.  (Then the inferred type parameter in the expression can be resolved from the ascription.)  So these are all the same and compile:
```
    let _: Foo = Foo::Baz;
    let _: Foo<String> = Foo::Baz;
    let _: Foo = Foo::<_>::Baz;
    let _: Foo<String> = Foo::<_>::Baz;
```

And then we have `<T>::` in an expression.  The `T` in `<T>` is always a type -- on its own inside the `<>`, `T` is not part of an expression.  Similar to type ascription, you can't elide non-defaulted parameters, but if you elide default parameters they take on the default value.  So these are the same:
```
    let _ = <Foo>::Baz;
    let _ = <Foo<String>>::Baz;
```

[`<T>` was introduced in RFC 0132](https://rust-lang.github.io/rfcs/0132-ufcs.html#paths-that-begin-with-a-type_segment), which also says that `Foo::<_>::...` is now equivalent to `<Foo<_>>::...`.  And sure enough, these don't compile either, as the inference still does not fallback to defaults.
```
    let _ = <Foo<_>>::Baz;
    let _: Foo<_> = Foo::Baz;
```

---

Summary
* In an expression position (where you need turbofish)
  * You can elide parameters whether they are defaulted or not
  * Elided parameters, with or without a default, will become fresh inference variables (`_`)
    * Which do not fallback to parameter defaults
* In a type position (where you don't need turbofish)
  * You cannot elide non-default parameters
  * Elided default parameters will use the default, not inference
  * "Type position" includes type ascription and `<T>`

**Comment by peter-kehl** [comment]
1. Is this still a tracking issue for https://github.com/rust-lang/rfcs/blob/master/text/0213-defaulted-type-params.md (as referenced from that RFC, and from https://doc.rust-lang.org/nightly/unstable-book/language-features/default-type-parameter-fallback.html)?
\
If so/As such, shouldn't this issue (or whatever tracking issue) be open (while the feature is, at least somewhat, on `nightly`) - regardless of unclear future for it?

2. Is this (or any other `nightly` feature, or another RFC) supposed to apply to function return types, where the return type is not related to any of the arguments (or no arguments at all)? (I can't see anything else relevant in the Unstable Book, nor at https://rust-lang.github.io/rfcs).
\
(The following currently fails even with `#![feature(default_type_parameter_fallback)]`: `cannot infer type of the type parameter A...`).
```rust
pub fn empty_vec<A=u32>() -> Vec<A> {
    Vec::new()
}

pub fn use_empty_vec() {
    empty_vec();
}
```

**Comment by traviscross** [comment]
Regarding the motivation to solve this issue, here's an example of code using the common builder pattern that seems like it should compile, but does not because of this.

The ergonomics of forcing the caller to specify the types of the fields not set are just horrible.  It entirely defeats the purpose of using the builder pattern.  So the only practical workaround is to use boxing and dynamic dispatch (which itself is not always possible because of object safety).

```rust
use std::future::Future;

type DefaultFuture = std::future::Pending<()>;

pub struct Builder<F1 = DefaultFuture, F2 = DefaultFuture> {
    fut1: Option<F1>,
    fut2: Option<F2>,
}

impl<F1, F2> Builder<F1, F2>
where
  F1: Future<Output = ()>,
  F2: Future<Output = ()>,
{
    pub fn new() -> Self {
        Self { fut1: None, fut2: None }
    }

    pub fn with_fut1(mut self, f: F1) -> Self {
        self.fut1 = Some(f);
        self
    }

    pub fn with_fut2(mut self, f: F2) -> Self {
        self.fut2 = Some(f);
        self
    }

    pub fn build(self) -> () {
        todo!()
    }
}

fn main() {
    let _foo = Builder::new()
        //.with_fut1(async move {}) // It should be optional to set this.
        .with_fut2(async move {})
        .build();
    // But it's not...
    // Error: cannot infer type of the type parameter `F1` declared on the struct `Builder`
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9e066c548aec7badac7f8791aee18eca)

As @peter-kehl suggested, until there is a "de-RFC" approved (e.g. as there was for [type ascription](https://rust-lang.github.io/rfcs/3307-de-rfc-type-ascription.html)) that cancels [RFC 213](https://rust-lang.github.io/rfcs/0213-defaulted-type-params.html) (and hopefully would document the problems that need to be overcome), it seems as though this tracking issue should remain open.

**Comment by peter-kehl** [comment]
Someone with access, or @Gankra or @jroesch, could you update the very first comment (issue description) to reflect the following, so as to save readers' time, please.

The current "EDIT" part (updated in July 2019) is increasing confusion, because
- in one paragraph it mentions two parts:
-- "stalled on disagreements... problem found during implementation" and 
-- it links to [See the internals thread where this was detailed and discussed](https://internals.rust-lang.org/t/interaction-of-user-defined-and-integral-fallbacks-with-inference/2496).
- (due to GitHub's design) the reader doesn't see right away that the "EDIT" part was added in 2019, two years after that (referred) internals thread was closed.

Unfortunately, that internals thread was auto-closed, and the last comment (from September 2015) says that "the lang subteam" has had a "consensus". So, it doesn't explain the problem mentioned in the "EDIT" in 2019.

As is, the "EDIT" paragraph easily confuses readers to search for the problem in that internals thread. Please split into two paragraphs, or add some wording to make the separation clear.

Curious: Was the problem found during implementation only when there is a conflict between the user-supplied default type and an inferred type? Because I have a huge piece of work (for `alloc`) that would not have such a conflict (or, in internal or `nightly` code only).

**Comment by RalfJung** [comment]
There's still a feature gate, `default_type_parameter_fallback`, pointing to this issue. @rust-lang/types should the issue be reopened, the feature gate removed, or the feature gate be changed to point to a different issue?

**Comment by lcnr** [comment]
I think we should completely remove the feature gate. It's currently only partially implemented and requires some significant design work before we re-add it afaict.

**Comment by RalfJung** [comment]
Sadly that's not really possible currently, `typemap` uses this feature...

**Comment by RalfJung** [comment]
I will reopen the issue then -- as long as the feature gate exists, the tracking issue it points to should remain open.
